src/main/java/org/springframework/data/couchbase/repository/query/CouchbaseQueryExecution.java

```diff
@@ -0,0 +1,157 @@
+/*
+ * Copyright 2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.data.couchbase.repository.query;
+
+import java.util.List;
+
+import org.springframework.core.convert.converter.Converter;
+import org.springframework.data.couchbase.core.CouchbaseOperations;
+import org.springframework.data.couchbase.core.ExecutableFindByQueryOperation;
+import org.springframework.data.couchbase.core.query.Query;
+import org.springframework.data.domain.Pageable;
+import org.springframework.data.domain.Slice;
+import org.springframework.data.domain.SliceImpl;
+import org.springframework.data.repository.query.QueryMethod;
+import org.springframework.data.repository.support.PageableExecutionUtils;
+import org.springframework.util.Assert;
+
+/**
+ * Set of classes to contain query execution strategies. Depending (mostly) on the return type of a
+ * {@link org.springframework.data.repository.query.QueryMethod} a {@link Query} can be executed in various flavors.
+ *
+ * @author Michael Reiche
+ * @since 4.1
+ */
+@FunctionalInterface
+interface CouchbaseQueryExecution {
+
+	Object execute(Query query, Class<?> type, String collection);
+
+	/**
+	 * {@link CouchbaseQueryExecution} removing documents matching the query.
+	 */
+
+	final class DeleteExecution implements CouchbaseQueryExecution {
+
+		private final CouchbaseOperations operations;
+		private final QueryMethod method;
+
+		public DeleteExecution(CouchbaseOperations operations, QueryMethod method) {
+			this.operations = operations;
+			this.method = method;
+		}
+
+		/*
+		 * (non-Javadoc)
+		 * @see org.springframework.data.couchbase.repository.query.AbstractCouchbaseQuery.Execution#execute(org.springframework.data.couchbase.core.query.Query, java.lang.Class, java.lang.String)
+		 */
+		@Override
+		public Object execute(Query query, Class<?> type, String collection) {
+			return operations.removeByQuery(type).matching(query).all();
```

Comment:
the collection string here is not handled at all - should we fail explicitly if it is present or even better handle it?

Reply:
As above - we are going to handle it, but I wanted to keep that out of the initial branch.